program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,371 @@ use crate::c_oracle_header::{
+                    update_time,
+                )?;
+            }
+            _ => {
+                //invalidate all the entries in your way
+                //this is ok because THRESHOLD < Granuality
+                self.invaldate_following_entries(num_skipped_entries);
+                self.max_update_time = update_time;
+            }
+        }
+        Ok(())
+    }
+}
+
+
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+/// Represents an Tick Tracker that has NUM_ENTRIES entries
+/// each tracking the last tick before every GRANUALITY seconds.
+pub struct TickTracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    prices:         [SignedTrackerRunningSum; NUM_ENTRIES], /* price running time
+                                                             * weighted sums */
+    confidences:    [UnsignedTrackerRunningSum; NUM_ENTRIES], /* confidence running
+                                                               * time
+                                                               * weighted sums */
+    current_entry:  usize,             //the current entry
+    entry_validity: [u8; NUM_ENTRIES], //entry validity
+}
+
+
+impl<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64>
+    TickTracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+{
+    ///invalidates current_entry, and increments self.current_entry num_entries times
+    fn invaldate_following_entries(&mut self, num_entries: usize) {
+        for _ in 0..num_entries {
+            self.entry_validity[self.current_entry] = 0;
+            self.current_entry = Self::get_next_entry(self.current_entry);
+        }
+    }
+}
+
+
+impl<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64>
+    Tracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+    for TickTracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+{
+    fn initialize(&mut self) -> Result<(), OracleError> {
+        Ok(())
+    }
+
+    ///add a new price to the tracker
+    fn add_price(
+        &mut self,
+        prev_time: i64,
+        _prev_price: i64,
+        _prev_conf: u64,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError> {
+        let num_skipped_entries = Self::get_num_skipped_entries_capped(prev_time, current_time)?;
+        match num_skipped_entries {
+            0 => {
+                self.prices[self.current_entry] = current_price;
+                self.confidences[self.current_entry] = current_conf;
+            }
+            1 => {
+                self.entry_validity[self.current_entry] =
+                    if Self::get_time_to_entry_end(prev_time) >= THRESHOLD {
+                        0
+                    } else {
+                        1
+                    };
+                self.current_entry += 1;
+                self.prices[self.current_entry] = current_price;
+                self.confidences[self.current_entry] = current_conf;
+            }
+            _ => {
```

Comment:
This case is more subtle right. Because before it went down for many entries, maybe that price was close enough to the boundary.

Reply:
I think the if statement on 311 catches this case, right?

Reply:
Yep, my comment is outdated